In [2]:
import sys
import os

In [3]:
project_dir = os.path.dirname(os.getcwd())

In [4]:
sys.path.append(os.path.abspath(os.path.join(os.path.dirname("classifier.ipynb"), "..")))

## Import the Classes

In [6]:
from models.decisiontreeregressor import DecisionTreeRegressor
from models.randomforestregressor import RandomForestRegressor
from models.xgboostregressor import XGBoostRegressor
from models.linearregression import LinearRegression
from models.linearregression_closed import LinearRegressionClosed

In [7]:
import pandas as pd
import numpy as np

## Load the data

In [8]:
insurance = pd.read_csv(os.path.join(project_dir, "data/insurance.csv"))

In [9]:
X = insurance.drop(columns="target")
y = insurance["target"]

KeyError: "['target'] not found in axis"

In [ ]:
n = len(X)

In [ ]:
indices = np.arange(n)
np.random.shuffle(indices)

In [ ]:
s1 = int(0.75 * n)

In [ ]:
X_train = X.iloc[indices[:s1],]
y_train = y.iloc[indices[:s1],]
X_test = X.iloc[indices[s1:],]
y_test = y.iloc[indices[s1:],]

## Logistic Regression

In [ ]:
lg = LogisticRegression()

In [ ]:
lg.fit(X_train, y_train)

/Users/vallurileelasaikrishna/Documents/from-scratch/models/logisticregression.py:16: RuntimeWarning: overflow encountered in exp
  return 1 / (1 + np.exp(-z))


In [ ]:
lg.accuracy(X_test, y_test)

/Users/vallurileelasaikrishna/Documents/from-scratch/.venv/lib/python3.13/site-packages/pandas/core/arraylike.py:402: RuntimeWarning: overflow encountered in exp
  result = getattr(ufunc, method)(*inputs, **kwargs)


np.float64(0.5758754863813229)

## Decision Tree

In [ ]:
dtc = DecisionTreeClassifier()

In [ ]:
dtc.fit(X_train, y_train)

In [ ]:
dtc.accuracy_test(X_test, y_test)

np.float64(0.8871595330739299)

## Random Forest

In [ ]:
rfc = RandomForestClassifier()

In [ ]:
rfc.fit(X_train, y_train)

In [ ]:
rfc.accuaracy_test(X_test, y_test)

np.float64(0.9260700389105059)

## XGBoost

In [ ]:
xgbc = XGBoostClassifier()

xgb takes train and val at the same time for fit

In [ ]:
s1 = int(0.7 * n)
s2 = int(0.85 * n)

In [ ]:
X_train = X.iloc[indices[:s1],]
y_train = y.iloc[indices[:s1],]
X_val = X.iloc[indices[s1:s2],]
y_val = y.iloc[indices[s1:s2],]
X_test = X.iloc[indices[s2:],]
y_test = y.iloc[indices[s2:],]

In [ ]:
xgbc.fit(X_train, y_train, X_val, y_val)

Early stopping at tree 584


In [ ]:
xgbc.accuracy(X_test, y_test)

np.float64(0.9514563106796117)

### Model Performance Comparison

| Model | Accuracy |
|--------|----------:|
| Logistic Regression | **0.5759** (57.59%) |
| Decision Tree | **0.8872** (88.72%) |
| Random Forest | **0.9261** (92.61%) |
| XGBoost | **0.9515** (95.15%) |